src/env.rs

```diff
@@ -4,65 +4,90 @@ use crate::abi_impl::{abi_bail, get_memory, ABIResult};
 use crate::types::Interface;
 use anyhow::Result;
 use as_ffi_bindings::{Read, StringPtr};
-use wasmer::{HostEnvInitError, Instance, WasmerEnv};
-use wasmer_middlewares::metering::{self, set_remaining_points, MeteringPoints};
-
-/// Error that append when a smartcontract try to call massa-std outside the
-/// main function. Wasmer hasn't the time to call `fn init_with_instance`
-pub const EXEC_INSTANCE_ERR: &str = "Unable to create execution instance. Make
-sure that you are not trying to call massa-std outside the main function";
+use wasmer::{Global, HostEnvInitError, Instance, WasmerEnv};
 
 #[derive(Clone)]
 pub struct Env {
     pub wasm_env: as_ffi_bindings::Env,
     pub interface: Box<dyn Interface>,
-    pub instance: Option<Instance>,
+    pub remaining_points: Option<Global>,
+    pub exhausted_points: Option<Global>,
 }
 
 impl Env {
     pub fn new(interface: &dyn Interface) -> Env {
         Env {
             wasm_env: Default::default(),
             interface: interface.clone_box(),
-            instance: None,
+            remaining_points: None,
+            exhausted_points: None,
         }
     }
 }
 
 impl WasmerEnv for Env {
     fn init_with_instance(&mut self, instance: &Instance) -> Result<(), HostEnvInitError> {
         self.wasm_env.init_with_instance(instance)?;
-        self.instance = Some(instance.clone());
+        self.remaining_points = Some(
+            instance
+                .exports
+                .get_with_generics_weak("wasmer_metering_remaining_points")
+                .map_err(HostEnvInitError::from)?,
+        );
+        self.exhausted_points = Some(
+            instance
+                .exports
+                .get_with_generics_weak("wasmer_metering_points_exhausted")
+                .map_err(HostEnvInitError::from)?,
+        );
         Ok(())
     }
 }
 
-pub fn get_remaining_points_for_env(env: &Env) -> ABIResult<u64> {
-    let instance = match env.instance.clone() {
-        Some(instance) => instance,
-        None => abi_bail!(EXEC_INSTANCE_ERR),
+/// Get remaining metering points
```

Comment:
I think it's also a good idea to add links to the wasmer metering middleware logic, since we're replicating that logic in here as opposed to calling their functions.

Reply:
Done.